## Step 1: Installation and start spark session

In [1]:
pip install pyspark


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
!pip install numpy pandas

Defaulting to user installation because normal site-packages is not writeable


Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.


In [1]:
import numpy as np
import pandas as pd

print(np.__version__)
print(pd.__version__)


1.24.4
2.0.3


In [3]:
from pyspark.sql import SparkSession

# Stop any existing Spark session first
try:
    spark.stop()
except:
    pass

# Create a clean Spark session
spark = SparkSession.builder \
    .appName("AI Impact Analysis") \
    .master("local[*]") \
    .getOrCreate()

print(spark.version)


3.5.7


## Step 2: Manual Data Splitting Across Nodes
Split the dataset into three parts: one for master, one for worker 1, and one for worker 2.


In [19]:
import pandas as pd


full_df = pd.read_csv("AI_Impact_on_Jobs_2030.csv")

df_master = full_df.sample(frac=0.34, random_state=42)
remaining = full_df.drop(df_master.index)
df_worker1 = remaining.sample(frac=0.5, random_state=42)
df_worker2 = remaining.drop(df_worker1.index)

df_master.to_csv("AI_Impact_on_Jobs_2030_master.csv", index=False)
df_worker1.to_csv("AI_Impact_on_Jobs_2030_worker1.csv", index=False)
df_worker2.to_csv("AI_Impact_on_Jobs_2030_worker2.csv", index=False)

print("Datasets split successfully:")
print("Master:", len(df_master), "rows")
print("Worker1:", len(df_worker1), "rows")
print("Worker2:", len(df_worker2), "rows")


Datasets split successfully:
Master: 1020 rows
Worker1: 990 rows
Worker2: 990 rows


## Step 3: Load Dataset on Each Node
Load the CSV partitions into Spark DataFrames simulating master and workers.


In [4]:
from pyspark.sql import SparkSession

spark = (
    SparkSession.builder
    .appName("AI Impact Analysis")
    .master("local[*]")  # local[*] works in notebook for testing
    .getOrCreate()
)


df_master = spark.read.csv("AI_Impact_on_Jobs_2030_master.csv", header=True, inferSchema=True)
df_worker1 = spark.read.csv("AI_Impact_on_Jobs_2030_worker1.csv", header=True, inferSchema=True)
df_worker2 = spark.read.csv("AI_Impact_on_Jobs_2030_worker2.csv", header=True, inferSchema=True)


df_all = df_master.union(df_worker1).union(df_worker2)

# Preview combined DataFrame
df_all.printSchema()
df_all.show(5)
print("Total rows combined:", df_all.count())


root
 |-- Job_Title: string (nullable = true)
 |-- Average_Salary: integer (nullable = true)
 |-- Years_Experience: integer (nullable = true)
 |-- Education_Level: string (nullable = true)
 |-- AI_Exposure_Index: double (nullable = true)
 |-- Tech_Growth_Factor: double (nullable = true)
 |-- Automation_Probability_2030: double (nullable = true)
 |-- Risk_Category: string (nullable = true)
 |-- Skill_1: double (nullable = true)
 |-- Skill_2: double (nullable = true)
 |-- Skill_3: double (nullable = true)
 |-- Skill_4: double (nullable = true)
 |-- Skill_5: double (nullable = true)
 |-- Skill_6: double (nullable = true)
 |-- Skill_7: double (nullable = true)
 |-- Skill_8: double (nullable = true)
 |-- Skill_9: double (nullable = true)
 |-- Skill_10: double (nullable = true)

+-------------------+--------------+----------------+---------------+-----------------+------------------+---------------------------+-------------+-------+-------+-------+-------+-------+-------+-------+-------+----

## Step 4: Data Partitioning and Spark SQL Processing
Repartition the data, clean it, and create features using SQL queries.


In [5]:
#  Create temp view for SQL queries
df_all.createOrReplaceTempView("ai_jobs")

spark.sql("""
SELECT AI_Exposure_Index, COUNT(*) AS cnt
FROM ai_jobs
GROUP BY AI_Exposure_Index
ORDER BY AI_Exposure_Index
""").show(20, truncate=False)

# Repartition by AI_Exposure_Index for parallel processing
df_all = df_all.repartition(8, "AI_Exposure_Index").cache()

#  Drop rows with nulls in columns used for ML
essential_cols = [
    "Risk_Category", "Education_Level", "Job_Title", "Average_Salary",
    "Years_Experience", "AI_Exposure_Index", "Tech_Growth_Factor",
    "Automation_Probability_2030"
] + [f"Skill_{i}" for i in range(1, 11)]
df_all = df_all.dropna(subset=essential_cols)

total_rows = df_all.count()
print("Total rows after repartitioning and cleaning:", total_rows)

df_all.show(5, truncate=False)


+-----------------+---+
|AI_Exposure_Index|cnt|
+-----------------+---+
|0.0              |12 |
|0.01             |38 |
|0.02             |28 |
|0.03             |28 |
|0.04             |28 |
|0.05             |26 |
|0.06             |30 |
|0.07             |30 |
|0.08             |27 |
|0.09             |27 |
|0.1              |33 |
|0.11             |27 |
|0.12             |28 |
|0.13             |27 |
|0.14             |27 |
|0.15             |24 |
|0.16             |27 |
|0.17             |32 |
|0.18             |32 |
|0.19             |40 |
+-----------------+---+
only showing top 20 rows

Total rows after repartitioning and cleaning: 3000
+-----------------+--------------+----------------+---------------+-----------------+------------------+---------------------------+-------------+-------+-------+-------+-------+-------+-------+-------+-------+-------+--------+
|Job_Title        |Average_Salary|Years_Experience|Education_Level|AI_Exposure_Index|Tech_Growth_Factor|Automation_Prob

## Step 5: Machine Learning Pipeline
Build a Spark MLlib pipeline with feature assembly, scaling, and logistic regression.


In [6]:
import time
from pyspark.ml import Pipeline
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

# -------------------------------
# Feature & Label Preparation
# -------------------------------

label_indexer = StringIndexer(
    inputCol="Risk_Category",
    outputCol="label",
    handleInvalid="keep"
)

education_indexer = StringIndexer(
    inputCol="Education_Level",
    outputCol="education_idx",
    handleInvalid="keep"
)

job_indexer = StringIndexer(
    inputCol="Job_Title",
    outputCol="job_idx",
    handleInvalid="keep"
)

encoder = OneHotEncoder(
    inputCols=["education_idx", "job_idx"],
    outputCols=["education_vec", "job_vec"]
)

numeric_features = [
    "Average_Salary",
    "Years_Experience",
    "AI_Exposure_Index",
    "Tech_Growth_Factor",
    "Automation_Probability_2030"
] + [f"Skill_{i}" for i in range(1, 11)]

assembler = VectorAssembler(
    inputCols=numeric_features + ["education_vec", "job_vec"],
    outputCol="features"
)

lr = LogisticRegression(
    featuresCol="features",
    labelCol="label",
    maxIter=50
)

pipeline = Pipeline(stages=[
    label_indexer,
    education_indexer,
    job_indexer,
    encoder,
    assembler,
    lr
])

# -------------------------------
# Train-Test Split
# -------------------------------

train_df, test_df = df_all.randomSplit([0.8, 0.2], seed=42)

# -------------------------------
# Measure Training Time
# -------------------------------

start_time = time.time()

model = pipeline.fit(train_df)

end_time = time.time()
training_time = end_time - start_time

# -------------------------------
# Prediction & Evaluation
# -------------------------------

pred = model.transform(test_df)

evaluator = MulticlassClassificationEvaluator(
    labelCol="label",
    predictionCol="prediction",
    metricName="f1"
)

print(f"Distributed Model Training Time: {training_time:.2f} seconds")
print("GLOBAL MODEL F1:", evaluator.evaluate(pred))


Distributed Model Training Time: 16.78 seconds
GLOBAL MODEL F1: 0.995045616843293


### Observation
The distributed machine learning model trained on the unified dataset achieved a very high F1-score of **0.995**, indicating excellent classification performance.  
Although the training time was **16.78 seconds**, it is expected due to distributed coordination and data shuffling across nodes, which trades higher computation time for better overall model accuracy.


### Step 6: Resource Management & Monitoring

- Spark executor memory, cores, and shuffle partitions can be configured for optimized performance.
- In local mode, the default settings are used, but on a real cluster you can adjust:
  - `spark.executor.memory` for executor RAM
  - `spark.executor.cores` for CPU cores per executor
  - `spark.sql.shuffle.partitions` for shuffle parallelism
- Monitor resource usage via the **Spark Web UI** at `http://localhost:4040`.




### Step 7: Performance Monitoring
- Monitor Spark cluster performance using the **Spark Web UI** at `http://localhost:4040`.
- Key metrics to check:
  - **Task distribution** across worker nodes
  - **Memory usage** per executor
  - **CPU utilization** per node
- Helps identify bottlenecks and optimize resource allocation.

### Step 8: Tuning the Machine Learning Model
- Fine-tune the ML model by adjusting **hyperparameters** (e.g., regularization, maxIter, tree depth).
- Use **cross-validation** to evaluate different configurations and select the best model.
- Goal: Improve model accuracy, precision, or F1-score on distributed data.

### Part 2 — Step 1: Simultaneous ML Jobs on Each Node
- Train separate machine learning models on the Master and Worker node datasets.
- Each node handles its own portion of the dataset independently.


In [7]:
import time

# --- Part 2.1: Prepare Node-Specific Datasets ---

# Split df_all roughly into three parts to simulate nodes
master_df, worker1_df, worker2_df = df_all.randomSplit([0.33, 0.33, 0.34], seed=42)

# Verify counts
print("Master node dataset count:", master_df.count())
print("Worker 1 dataset count:", worker1_df.count())
print("Worker 2 dataset count:", worker2_df.count())

# --- Part 2.1: Train Separate ML Models on Each Node ---

# Function to train ML pipeline on a node's dataset
def train_node_model(node_df, node_name):
    # Train/Test split
    train_df, test_df = node_df.randomSplit([0.8, 0.2], seed=42)
    
    # ⏱️ Start timing
    start_time = time.time()
    
    # Fit the pipeline
    model = pipeline.fit(train_df)
    
    # ⏱️ End timing
    end_time = time.time()
    training_time = end_time - start_time
    
    # Predictions
    pred = model.transform(test_df)
    
    # Evaluate F1
    f1_score = evaluator.evaluate(pred)
    
    print(f"{node_name} Training Time: {training_time:.2f} seconds")
    print(f"F1 Score for {node_name}: {f1_score:.4f}\n")
    
    return model, training_time

# Train models on each node
model_master, time_master = train_node_model(master_df, "Master Node")
model_worker1, time_worker1 = train_node_model(worker1_df, "Worker 1")
model_worker2, time_worker2 = train_node_model(worker2_df, "Worker 2")


Master node dataset count: 1006
Worker 1 dataset count: 965
Worker 2 dataset count: 1029
Master Node Training Time: 9.43 seconds
F1 Score for Master Node: 0.9572

Worker 1 Training Time: 8.77 seconds
F1 Score for Worker 1: 0.9898

Worker 2 Training Time: 6.48 seconds
F1 Score for Worker 2: 0.9766



### Observation
Each node trained its machine learning model on a smaller, local portion of the dataset, resulting in **shorter training times** compared to the distributed model.  
While training was faster (6–9 seconds), the F1-scores were **slightly lower and varied across nodes**, showing that models trained on partial data may lose some generalization capability compared to a unified distributed model.


### Part 2 — Step 2: Distributed Machine Learning on Unified Dataset
- Train a single machine learning model on the unified dataset (`df_all`) spread across all nodes.
- This simulates a distributed ML job using the entire dataset.


In [8]:
import time

# --- Part 2.2: Distributed ML on Unified Dataset ---

# Train/Test split on the unified dataset
train_df, test_df = df_all.randomSplit([0.8, 0.2], seed=42)

# ⏱️ Start timing
start_time = time.time()

# Fit the same pipeline
distributed_model = pipeline.fit(train_df)

# ⏱️ End timing
end_time = time.time()
distributed_training_time = end_time - start_time

# Predictions
pred_dist = distributed_model.transform(test_df)

# Evaluate F1 score
f1_dist = evaluator.evaluate(pred_dist)

print(f"Distributed Model Training Time: {distributed_training_time:.2f} seconds")
print(f"F1 Score for Distributed Model: {f1_dist:.4f}")


Distributed Model Training Time: 7.61 seconds
F1 Score for Distributed Model: 0.9950


### Observation
The distributed machine learning model trained on the unified dataset achieved a **high F1-score of 0.995**, demonstrating strong predictive performance.  
The training time of **7.61 seconds** shows that distributed processing efficiently utilized cluster resources to train the model on the complete dataset.


### Part 2 — Step 3: Performance Monitoring & Comparison
- Monitor and compare model performance across separate node models and the distributed model.
- Compare F1 scores and dataset counts to analyze differences.


In [9]:
# --- Part 2.3: Compare Models ---

print("Dataset counts for each node and unified dataset:")
print(f"Master Node: {master_df.count()}")
print(f"Worker 1: {worker1_df.count()}")
print(f"Worker 2: {worker2_df.count()}")
print(f"Unified Dataset: {df_all.count()}")

print("\nF1 Scores Comparison:")
print(f"Master Node: {evaluator.evaluate(model_master.transform(master_df)):.4f}")
print(f"Worker 1: {evaluator.evaluate(model_worker1.transform(worker1_df)):.4f}")
print(f"Worker 2: {evaluator.evaluate(model_worker2.transform(worker2_df)):.4f}")
print(f"Distributed Model: {f1_dist:.4f}")

print("\nTraining Time Comparison (seconds):")
print(f"Master Node: {time_master:.2f}")
print(f"Worker 1: {time_worker1:.2f}")
print(f"Worker 2: {time_worker2:.2f}")
print(f"Distributed Model: {distributed_training_time:.2f}")


Dataset counts for each node and unified dataset:
Master Node: 1006
Worker 1: 965
Worker 2: 1029
Unified Dataset: 3000

F1 Scores Comparison:
Master Node: 0.9911
Worker 1: 0.9979
Worker 2: 0.9951
Distributed Model: 0.9950

Training Time Comparison (seconds):
Master Node: 9.43
Worker 1: 8.77
Worker 2: 6.48
Distributed Model: 7.61


### Observation
The unified distributed dataset contains all **3000 records**, while individual nodes process smaller subsets ranging from **965 to 1029 records**.  
Node-level models trained faster due to reduced data size but showed **variation in F1-scores**, whereas the distributed model achieved a **consistently high F1-score (0.995)** with balanced training time.  
This highlights the trade-off between **faster local training** and **better generalization** achieved through distributed learning.
